File: 2022/deno-aoc-2022/src/day-06.ts
function allDifferent(s: string, index: number, count: number): boolean {
  const set = new Set();
  for (let i = index; i < index + count; i++) {
    set.add(s[i]);
  }
  return set.size == count;
}

const day06a = async () => {
  const inputFile = await Deno.readTextFile("input/day-06.txt");
  const count = 4;

  for (let i = 0; i < inputFile.length - count; i++) {
    if (allDifferent(inputFile, i, count)) {
      console.log(i + count);
      return;
    }
  }
};

const day06b = async () => {
  const inputFile = await Deno.readTextFile("input/day-06.txt");
  const count = 14;

  for (let i = 0; i < inputFile.length - count; i++) {
    if (allDifferent(inputFile, i, count)) {
      console.log(i + count);
      return;
    }
  }
};

if (import.meta.main) {
  await day06a();
  await day06b();
}

File: 2023/swift-aoc-2023/Sources/day06.swift
import Foundation
import Parsing

struct Race {
    var time: Int
    var record: Int
}

func readInputA() -> [Race]? {
    if let input = try? String(contentsOfFile: "input/day-06.txt") {
        let parser = Parse(input: Substring.self) {
            "Time:"
            Many {
                Whitespace()
                Digits()
            }
            Whitespace()
            "Distance:"
            Many {
                Whitespace()
                Digits()
            }
        }

        do {
            let data = try parser.parse(input)
            return zip(data.0, data.1).map { Race(time: $0.0, record: $0.1) }
        } catch {
            print(error)
        }
    } else {
        print("Couldn't open file")
    }

    return []
}

func readInputB() -> Race? {
    if let input = try? String(contentsOfFile: "input/day-06.txt") {
        let parser = Parse(input: Substring.self) {
            "Time:"
            Many {
                Whitespace()
                Digits()
            }
            Whitespace()
            "Distance:"
            Many {
                Whitespace()
                Digits()
            }
        }

        do {
            let data = try parser.parse(input)
            if let time = Int(data.0.reduce("") { $0 + String($1) }),
               let record = Int(data.1.reduce("") { $0 + String($1) })
            {
                return Race(time: time, record: record)
            }
        } catch {
            print(error)
        }
    } else {
        print("Couldn't open file")
    }

    return nil
}

func day06a() {
    if let races = readInputA() {
        var results: [Int] = []
        for race in races {
            var count = 0
            for x in 0 ... race.time {
                let d = (race.time - x) * x
                if d > race.record {
                    count += 1
                }
            }
            results.append(count)
        }
        print(results.reduce(1, *))
    }
}

func day06b() {
    if let race = readInputB() {
        var count = 0
        for x in 0 ... race.time {
            let d = (race.time - x) * x
            if d > race.record {
                count += 1
            }
        }
        print(count)
    }
}
